File: src/ultraloom/worktree.py
"""What git says has changed below a directory.

Its own module because two callers need the *same* answer: the CLI takes a
run's baseline when the run starts, and the verify flow's guard reads the tree
again after a repair pass. A second implementation of the same git call would
drift in exactly the parsing details this file exists for, and the two answers
would then be compared against each other.

Below the harness on purpose: it raises its own error and knows nothing about
flows, exit codes or runs, so `ultraloom check` may import it (spec 15.2).
"""

from __future__ import annotations

import subprocess
from pathlib import Path

# Where a run keeps its journal and its marker. Defined here and not in the CLI
# that writes them: this module has to know what to leave out of its answer, and
# it sits below the CLI, so the constant travels the other way round.
RUN_DIR = ".ultraloom/runs"


class WorktreeError(RuntimeError):
    """Raised when git cannot answer what changed. Never read as "nothing"."""


def changed_files(root: Path) -> tuple[str, ...]:
    """Every path git reports as changed, added or untracked below `root`.

    `status` and not `diff`, because a repairer may add a file, and an
    untracked file is invisible to `diff`. `-z` because a path holding
    non-ASCII comes back quoted otherwise, and `-uall` because the default
    collapses a whole untracked directory into one entry that is not a path to
    any file.

    Paths come back relative to `root`, which is not what git says: porcelain
    output is relative to the *repository* root whatever the working directory
    is, and there is no porcelain option that changes it. Where the two are the
    same directory the difference is invisible, which is why it went unnoticed
    -- but in a monorepo run with `--root package`, git answers
    "package/tests/test_x.py" while the project's own `[verify].tests` says
    "tests/", nothing a caller configured ever matches, and a guard built on
    this answer is silently off. Anything outside `root` is dropped for the
    same reason: it is not this project's change.

    A root git *ignores* is refused rather than answered. Such a directory is
    still inside the repository, so every call here succeeds and the arithmetic
    below is right -- but `status` never lists an ignored file, so the answer
    is empty however much changed. A copy of a project parked below an ignored
    path is the ordinary way to end up there, and read as "nothing changed" it
    switches the guard off while every run keeps reporting success.

    What `RUN_DIR` holds is dropped as well, and for a reason of its own: those
    files are ultraloom's, not the project's. Every run writes its journal and
    its marker while the repair agent works, so a guard reading this answer
    would report them as the agent's doing -- and a project that lists
    `.ultraloom/` among its protected paths would take exit 4 on every single
    run, named after files ultraloom wrote itself. The rest of `.ultraloom/`
    stays visible: `config.toml` holds the thresholds a check is measured
    against, and an agent editing that one is exactly what the guard is for.
    """
    # Before the question rather than after the answer: an ignored root is
    # wrong whatever comes back, and a change *elsewhere* in the repository
    # would otherwise carry the call past this point and be dropped by the
    # relocation below -- an empty answer again, and this time an unchecked one.
    _refuse_if_ignored(root)
    return _relocate(root, _parse_status(_status(root)))


def _relocate(root: Path, paths: tuple[str, ...]) -> tuple[str, ...]:
    """Repository-relative paths as a caller below `root` spells them.

    Anything outside `root` is dropped -- it is not this project's change --
    and so is everything below `RUN_DIR`, which is ultraloom's own doing and
    never the repairer's. Both callers go through here, because two spellings
    of the same path would end up being compared against each other.
    """
    if not paths:
        # Nothing to relocate, so the second git call is not worth its process.
        return paths
    prefix = _prefix(root)
    if prefix:
        paths = tuple(path[len(prefix) :] for path in paths if path.startswith(prefix))
    # After the relocation, so the comparison is against the spelling a caller
    # below `root` would use rather than the repository-relative one.
    return tuple(path for path in paths if not path.startswith(RUN_DIR + "/"))


def changed_since(root: Path, base: str) -> tuple[str, ...]:
    """Every path that differs between `base` and the working tree, below `root`.

    The union of two questions, because neither answers alone: `diff` sees what
    was committed since `base` but is blind to an untracked file, and `status`
    sees the untracked file but reads a committed change as a clean tree. That
    second blindness is why this function exists -- a repairer that commits its
    edit leaves `status` with nothing to report, and the guard built on it then
    lets an edited test file through.

    `--no-renames`, so a rename comes back as its old *and* its new path. Git
    would otherwise report the pair as one entry, and a test moved out of the
    way would be a path the guard never compares against its protected list.

    Content-based, so a `reset`, a `rebase` or an `amend` hides nothing: this
    compares the tree of `base` against the tree on disk, not two histories.

    An unresolvable `base` is a `WorktreeError` and never an empty answer, for
    the reason every refusal in this module has: a question git cannot answer
    must not be read as "nothing changed".
    """
    _refuse_if_ignored(root)
    diff = _git(root, "diff", "--name-only", "--no-renames", base)
    committed = tuple(line for line in diff.splitlines() if line)
    reported = _parse_status(_status(root))
    # One call over both answers rather than one per answer: `_relocate` asks
    # git for the prefix, and that process is the same for either spelling.
    return tuple(sorted(set(_relocate(root, committed + reported))))


def head_commit(root: Path) -> str:
    """The commit a run starts on, as git spells it.

    `rev-parse HEAD` and not `--short`: the answer travels in a run marker and
    is read back rounds later, and an abbreviated SHA is only unique for as
    long as the repository stays the size it was.

    Three ways of having no answer, all of them `WorktreeError`: no repository,
    a repository without a commit -- `git init` leaves HEAD naming a branch
    that does not exist yet -- and a root git ignores. The last one is why
    `_refuse_if_ignored` is asked here at all: such a directory *is* inside a
    repository, so `rev-parse` answers readily with the surrounding
    repository's HEAD. Measuring against that is worse than not measuring, as
    every file of the parked copy then reads as the repairer's doing.
    """
    _refuse_if_ignored(root)
    return _git(root, "rev-parse", "HEAD").strip()


def _run(root: Path, *arguments: str) -> subprocess.CompletedProcess[str]:
    """One git call below `root`, for the callers that read the return code."""
    try:
        return subprocess.run(
            ("git", *arguments),
            cwd=root,
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            check=False,
        )
    except OSError as error:
        # A directory that is not there never reaches a return code: the spawn
        # itself fails. Same answer as a non-zero one -- see `_git`.
        raise WorktreeError(f"cannot inspect the working tree in {root}: {error}") from error


def _refuse_if_ignored(root: Path) -> None:
    """Refuse a root git answers about but never answers with.

    `check-ignore` exits 1 for a path it does not ignore, which is the ordinary
    case and no failure at all -- so this call reads the return code itself
    rather than going through `_git`. Every other code is read as "not
    ignored" on purpose: the calls that follow refuse a directory git cannot
    answer about anyway, and turning the difference between exit 1 and exit 128
    into a second way of failing here would only make that refusal less clear.
    """
    if _run(root, "check-ignore", "-q", ".").returncode == 0:
        raise WorktreeError(
            f"git ignores {root}, so it can never report a change there -- "
            "run ultraloom in a working tree of its own"
        )


def _git(root: Path, *arguments: str) -> str:
    """One git call below `root`, with both ways of having no answer refused."""
    result = _run(root, *arguments)
    if result.returncode != 0:
        # A caller that cannot see the working tree must not carry on as if it
        # had seen an empty one. Reading an unanswerable question as "nothing
        # changed" would disable exactly the rules this answer feeds.
        raise WorktreeError(f"cannot inspect the working tree in {root}: {result.stderr.strip()}")
    return result.stdout


def _status(root: Path) -> str:
    return _git(root, "status", "--porcelain", "-z", "-uall")


def _prefix(root: Path) -> str:
    """How deep `root` sits below the repository root, as git spells it.

    Empty when the two are the same directory, and always slash-terminated
    otherwise -- so it is exactly the string to cut off the front of a
    repository-relative path.
    """
    return _git(root, "rev-parse", "--show-prefix").strip()


def _parse_status(output: str) -> tuple[str, ...]:
    """The paths out of a `--porcelain -z` answer, read field by field.

    Most fields are "XY path". A rename or a copy is the exception: git emits
    *two* fields for it, and only the first carries the three-character prefix
    -- the second is the original path, bare. Cutting three characters off that
    one too would turn "tests/test_cli.py" into "s/test_cli.py", and a test
    renamed out of the way would walk straight past the guard.
    """
    fields = [field for field in output.split("\0") if field]
    paths: list[str] = []
    index = 0
    while index < len(fields):
        field = fields[index]
        paths.append(field[3:])
        index += 1
        if field[:1] in ("R", "C") and index < len(fields):
            paths.append(fields[index])
            index += 1
    return tuple(paths)

File: cmd/guard/worktree.go
// The worktree side of the guard: what a working tree cannot own itself.
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/xidus90/ultra-loom/internal/gitenv"
	"github.com/xidus90/ultra-loom/internal/junction"
	"github.com/xidus90/ultra-loom/internal/mirrorcfg"
	"github.com/xidus90/ultra-loom/internal/sessions"
	"github.com/xidus90/ultra-loom/internal/worktreetopo"
)

// The two prefixes a stored reparse target can start with. `\??\` is the NT
// object-manager form, it is not a path Go opens, and it is the only one
// measured here: on 2026-09-07 both junction.Create and `mklink /J` stored
// it. `\\?\` is taken off as well so that both forms compare alike -- and not
// because anything here was seen to store it. Neither prefix identifies
// anything: the reparse tag does that, and junction.Target has checked it.
var ntPrefixes = []string{`\??\`, `\\?\`}

// runWorktreeLink puts the configured directories in place and sweeps what is
// left of gone worktrees.
//
// Silent on success, because it runs at every session start in every project
// on the machine. The only thing it reports is a junction that was needed and
// could not be made -- and that one it reports loudly, since the directory it
// stands for may be the pinned runtime every other hook needs.
//
// stdout is taken and not written to: every answer this subcommand has is
// either silence or a fault, and a fault belongs on stderr.
func runWorktreeLink(stdout, stderr io.Writer, root string) int {
	topology, err := worktreetopo.Read(root)
	if err != nil {
		// Every failure of that call is the same answer, and the answer is
		// "nothing to do": worktreetopo.Read wraps ErrNoRepository around each
		// of its two error returns rather than telling a broken git from an
		// unrepositoried directory. Distinguishing them here would be a branch
		// nothing can enter.
		return ExitOK
	}
	mirror, err := mirrorcfg.Mirror(topology.Main)
	if err != nil {
		fmt.Fprintf(stderr, "ultraloom-guard worktree-link: %v\n", err)
		return ExitInternal
	}
	if len(mirror) == 0 {
		return ExitOK
	}

	code := ExitOK
	if topology.IsWorktree(root) {
		if err := link(root, topology.Main, mirror); err != nil {
			fmt.Fprintf(stderr, "ultraloom-guard worktree-link: %v\n", err)
			code = ExitInternal
		}
	}
	// The sweep runs whether or not this directory is a worktree: a session in
	// the main checkout is the ordinary way to notice that a worktree is gone.
	if err := sweep(topology, mirror); err != nil {
		fmt.Fprintf(stderr, "ultraloom-guard worktree-link: %v\n", err)
		code = ExitInternal
	}
	return code
}

// How long a session's state file counts for. Nothing deletes these files, so
// the mtime is the only liveness there is to read, and it is as good as the
// writes: session_start.py writes at session start (:59), stop.py on every
// block and every pass (:250, :283), and subagent_start.py on every subagent
// dispatch (:38) -- the last of those gated on the payload alone and on no
// configuration at all. So the file is as young as the last turn that ended,
// or the last subagent dispatched, and only a session that does neither ages
// past its start.
//
// A day, because the two errors are not the same size. Too long leaves a
// junction standing, and that costs nothing: it occupies no disk, `link` skips
// it at the next session start as already there, and `sweep` takes it out once
// the worktree is gone -- the same state `link` deliberately creates. Too
// short takes it out from under a live session or an open Godot editor reading
// 4.2 GB through it. A working day is also the unit in which a human answers
// "is that session still mine?".
//
// No number closes this hole, and this one does not either: the fix is a write
// on the live side -- worktree-link touching the file at session start, or a
// write from the SessionEnd side -- which is follow-up work and not this.
const sessionStale = 24 * time.Hour

// runWorktreeUnlink takes the junctions back out -- but only if this was the
// last session on the tree.
//
// CLAUDE.md documents sessions that share a checkout. Unlinking
// unconditionally would pull `.tools` out from under a session still running,
// or under an open Godot editor, and the 4.2 GB behind it is exactly what that
// editor is reading from.
//
// stdout is taken and not written to, for worktree-link's reason: every answer
// this subcommand has is either silence or a fault.
func runWorktreeUnlink(stdout, stderr io.Writer, stdin io.Reader, root string) int {
	var payload struct {
		SessionID string `json:"session_id"`
	}
	// A payload we cannot read is not a reason to remove anything: without an
	// id there is no way to tell our own state file from somebody else's, so
	// the count would always say "somebody else is here" -- and Forget would
	// be worse than useless, since safeName turns an empty id into "unnamed"
	// and that may be the collapsed name of a session that is still running.
	if err := json.NewDecoder(stdin).Decode(&payload); err != nil || payload.SessionID == "" {
		return ExitOK
	}

	topology, err := worktreetopo.Read(root)
	if err != nil {
		// Nothing to do, and the same collapse worktree-link makes: Read wraps
		// ErrNoRepository around each of its two error returns, so telling them
		// apart here would be a branch nothing can enter.
		return ExitOK
	}
	if !topology.IsWorktree(root) {
		return ExitOK
	}
	mirror, err := mirrorcfg.Mirror(topology.Main)
	if err != nil {
		fmt.Fprintf(stderr, "ultraloom-guard worktree-unlink: %v\n", err)
		return ExitInternal
	}
	if len(mirror) == 0 {
		return ExitOK
	}

	// Our own file goes before the count and before the early return below: it
	// has to be gone whether or not this was the last session, or the next run
	// finds it and reads a session that has ended as one still standing.
	if err := sessions.Forget(root, payload.SessionID); err != nil {
		fmt.Fprintf(stderr, "ultraloom-guard worktree-unlink: %v\n", err)
		return ExitInternal
	}
	others, err := sessions.Others(root, payload.SessionID, sessionStale)
	if err != nil {
		fmt.Fprintf(stderr, "ultraloom-guard worktree-unlink: %v\n", err)
		return ExitInternal
	}
	if others > 0 {
		return ExitOK
	}
	if err := unlink(root, topology.Main, mirror); err != nil {
		fmt.Fprintf(stderr, "ultraloom-guard worktree-unlink: %v\n", err)
		return ExitInternal
	}
	return ExitOK
}

// runWorktreeRemove is the safe way to get rid of a worktree.
//
// Measured on 2026-09-07 in a t.TempDir() fixture: `git worktree remove
// --force` on a worktree holding a junction exits 0 with no output, drops the
// porcelain entry, and leaves both the directory and the junction standing.
// The directory is then a tree git no longer knows, with a link into the main
// checkout still in it. So the junctions come out first, and git is asked
// afterwards.
//
// Two refusals, and both before anything is touched -- the read of the
// topology is the only thing that happens ahead of them, and it only asks git
// a question. The main checkout is refused first, because a wrapper whose
// worst outcome is deleting the repository has to say no to that one before
// anything else; a directory git holds no working tree at is refused because
// there is then nothing here to remove and every candidate is somebody's data.
// Both rest on os.SameFile identity and not on the spelling of the argument,
// so no relative form, trailing separator or link path gets past them.
//
// This one writes to stdout, unlike worktree-link and worktree-unlink: those
// fire at every session start and end in every project, this one is run by
// hand, and a person deleting something should read what was deleted.
func runWorktreeRemove(stdout, stderr io.Writer, target string) int {
	topology, err := worktreetopo.Read(target)
	if err != nil {
		// A fault here and not the hook commands' silent "nothing to do": this
		// directory was named by hand, and the name was wrong.
		fmt.Fprintf(stderr, "ultraloom-guard worktree-remove: %v\n", err)
		return ExitInternal
	}
	if sameDir(target, topology.Main) {
		fmt.Fprintf(stderr,
			"ultraloom-guard worktree-remove: %s is the main checkout\n", target)
		return ExitInternal
	}
	worktree := registeredAs(topology, target)
	if worktree == "" {
		fmt.Fprintf(stderr,
			"ultraloom-guard worktree-remove: git does not hold %s as a worktree\n", target)
		return ExitInternal
	}
	mirror, err := mirrorcfg.Mirror(topology.Main)
	if err != nil {
		// Reported rather than treated as "no mirror": which junctions are
		// ours is then unknown, and asking git while that is unknown is the
		// order this whole subcommand exists to avoid.
		fmt.Fprintf(stderr, "ultraloom-guard worktree-remove: %v\n", err)
		return ExitInternal
	}
	if err := unlink(worktree, topology.Main, mirror); err != nil {
		fmt.Fprintf(stderr, "ultraloom-guard worktree-remove: %v\n", err)
		return ExitInternal
	}
	command := exec.Command("git", "worktree", "remove", "--force", worktree)
	command.Dir = topology.Main
	// See gitenv: GIT_DIR and its relatives outrank command.Dir, so without
	// the strip the removal would be asked of whatever GIT_DIR names instead
	// of this repository.
	command.Env = gitenv.Environ()
	if out, err := command.CombinedOutput(); err != nil {
		// git's own words go through: it is the only one that knows why it
		// refused. The junctions are already out by then, and nothing here
		// puts them back -- the next session's worktree-link does that.
		fmt.Fprintf(stderr, "ultraloom-guard worktree-remove: git: %v (%s)\n", err, out)
		return ExitInternal
	}
	// git's exit code says nothing about the directory. Measured on
	// 2026-09-07: with a junction inside it that nothing here took out -- no
	// mirror configured -- `git worktree remove --force` exited 0, dropped the
	// porcelain entry and left the tree standing. That is the leftover this
	// whole subcommand exists to prevent, so it is not something to print
	// "removed" over.
	//
	// The message says what was observed and nothing about the cause: a
	// junction none of ours, an open handle, an ACL and a virus scanner all
	// leave exactly this behind, and os.Lstat tells the four apart in no way
	// at all. The operator is the one who can look.
	if _, err := os.Lstat(worktree); err == nil {
		fmt.Fprintf(stderr,
			"ultraloom-guard worktree-remove: git dropped its entry but %s still stands;"+
				" nothing here removed it\n", worktree)
		return ExitInternal
	}
	fmt.Fprintf(stdout, "removed %s\n", worktree)
	return ExitOK
}

// registeredAs answers with git's own spelling of `target`, or "" if git holds
// no working tree there.
//
// The caller passes that spelling on rather than its own argument, because
// the two do not have to name the same directory: its git call runs with its
// working directory in the main checkout, so a relative argument would resolve
// there, while sameDir resolves it against this process's own. Taking the path
// out of the porcelain closes the gap and needs no filepath.Abs -- whose only
// failure mode, a working directory that cannot be read, no test can reach.
//
// Main is in Worktrees as well, so this matches it too -- which is harmless
// only because the caller has already refused it above.
func registeredAs(topology worktreetopo.Topology, target string) string {
	for _, worktree := range topology.Worktrees {
		if sameDir(target, worktree) {
			return worktree
		}
	}
	return ""
}

// unlink removes the configured junctions from one worktree.
//
// A configured path that is a real directory is left alone: it was never ours.
// The target is checked as well -- a junction pointing somewhere else is
// somebody's own arrangement, and removing it would be the same overreach as
// removing a real directory.
//
// standsInside first, for the reason written at that function: without it the
// candidate is inside the worktree by spelling only, and a junction at
// `<worktree>/.ultraloom` would make this remove the main checkout's own
// `.ultraloom/vendor` -- the pinned runtime every other hook needs, taken out
// at session end by the mechanism that exists to put it there.
func unlink(worktree, main string, mirror []string) error {
	for _, relative := range mirror {
		if !standsInside(worktree, relative) {
			continue
		}
		path := filepath.Join(worktree, filepath.FromSlash(relative))
		target, err := junction.Target(path)
		if err != nil {
			return err
		}
		if target == "" || !leadsInto(target, main) {
			continue
		}
		if err := junction.Remove(path); err != nil {
			return err
		}
	}
	return nil
}

// link makes every configured path in `worktree` lead into `main`.
//
// A path the worktree already owns is left alone, and so is one the main
// checkout does not have: the first may be this tree's own build output, and
// the second is nothing to mirror. Only an *absent* path here is ours to fill.
//
// An Lstat that fails for some other reason than absence is not caught, and
// the property that makes that safe is that such a path always ends in a
// reported fault rather than a silent skip -- by one of two routes since
// parentsPlainOrAbsent arrived, not one: that function refuses the path when
// the unreadable component is an intermediate, and junction.Create's os.Mkdir
// refuses it when it is the candidate itself. Measured on 2026-09-07 with a worktree that may
// not gain a subdirectory -- os.Lstat of the absent child still answers
// IsNotExist there, and the Mkdir inside Create is what refuses. Mkdir
// refusing an occupied path is the same property from the other side: it is
// why nothing here can overwrite what already stands at the path.
//
// parentsPlainOrAbsent before the create, for the reason written at that
// function: without it "inside the worktree" is spelling only, and this is
// the create side of what standsInside does for the two remove sides. The
// refusal is loud and not a skip, because a mirror that was needed and could
// not be made is the one thing this subcommand reports.
func link(worktree, main string, mirror []string) error {
	for _, relative := range mirror {
		target := filepath.Join(main, filepath.FromSlash(relative))
		if info, err := os.Stat(target); err != nil || !info.IsDir() {
			continue
		}
		path := filepath.Join(worktree, filepath.FromSlash(relative))
		if _, err := os.Lstat(path); err == nil {
			continue
		}
		if blocker, ok := parentsPlainOrAbsent(worktree, relative); !ok {
			return fmt.Errorf(
				"not making %s: %s is not a plain directory as far as Lstat can see",
				path, blocker)
		}
		// The parent may be missing: a configured path can be more than one
		// level deep, and only its first level is necessarily something git
		// brought along.
		if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
			return fmt.Errorf("making room for %s: %w", path, err)
		}
		if err := junction.Create(path, target); err != nil {
			return fmt.Errorf("%s: %w", relative, err)
		}
	}
	return nil
}

// sweep removes our junctions from directories git no longer holds.
//
// What makes a junction ours is where it is and what it points at: a reparse
// point at a configured path inside an unregistered worktree directory,
// leading into the main checkout. All three conditions are load-bearing.
// CLAUDE.md records directories under `.claude/worktrees/` that share the main
// index and are not worktrees, so Orphans names live work as well -- a real
// directory there is somebody's data, and a junction pointing anywhere else is
// somebody's own arrangement.
//
// No ledger -- a list of junctions we made would be a second state that can
// drift, and after a `Remove-Item -Recurse` on a worktree it would be wrong
// immediately. The price is that the hand-made junctions in this repository's
// worktrees are adopted, which is right: they are indistinguishable from ours
// by target and place, and they were made for the same reason.
func sweep(topology worktreetopo.Topology, mirror []string) error {
	orphans, err := topology.Orphans()
	if err != nil {
		return err
	}
	for _, orphan := range orphans {
		for _, relative := range mirror {
			if !standsInside(orphan, relative) {
				continue
			}
			path := filepath.Join(orphan, filepath.FromSlash(relative))
			target, err := junction.Target(path)
			if err != nil {
				return err
			}
			if target == "" {
				// Not a link: somebody's directory, and none of our business.
				continue
			}
			if !leadsInto(target, topology.Main) {
				continue
			}
			if err := junction.Remove(path); err != nil {
				return err
			}
		}
	}
	return nil
}

// standsInside says whether a configured path really lies where its spelling
// says: every component between `dir` and the candidate itself a plain
// directory, and none of them a link. Both *removing* callers need it -- the
// sweep about an orphaned worktree directory, unlink about a live one. The
// create side has a rule of its own, at parentsPlainOrAbsent, because a
// missing component is nothing for link to refuse.
//
// Without it a caller establishes "inside `dir`" by spelling alone, and a
// path's spelling does not decide where it goes: an open with
// FILE_FLAG_OPEN_REPARSE_POINT keeps the *final* component from being
// followed and nothing else, so a junction at `<dir>/.ultraloom` makes
// `<dir>/.ultraloom/vendor` read and remove the reparse point of
// `<main>/.ultraloom/vendor` -- the pinned runtime, taken out by the very
// mechanism that exists to put it there.
//
// Mode().IsDir() is the test, and it is false for a junction under both
// GODEBUG settings this module can be built with; the measurement behind that
// is recorded at junction.go's Target. The last component is left out because
// junction.Target is what decides about that one, and being a link is exactly
// what qualifies it.
//
// A component that cannot be stat'ed counts as not a plain directory. The safe
// direction is to leave a candidate alone: a junction skipped costs a stale
// directory nobody deletes, and the other way costs somebody's data.
func standsInside(dir, relative string) bool {
	// mirrorcfg hands out cleaned, slash-separated paths, so this split has no
	// empty component and no trailing one.
	components := strings.Split(relative, "/")
	path := dir
	for _, component := range components[:len(components)-1] {
		path = filepath.Join(path, component)
		info, err := os.Lstat(path)
		if err != nil || !info.Mode().IsDir() {
			return false
		}
	}
	return true
}

// parentsPlainOrAbsent says whether every component between `dir` and the
// candidate itself is either a plain directory or not there at all, and names
// the first one that is neither.
//
// This is what the create side needs, and standsInside is not it: standsInside
// wants each of those components to *exist* as a plain directory, while link
// legitimately fills missing ones in -- a configured path can be more than one
// level deep, and only its first level is necessarily something git brought
// along. Absent is therefore allowed here and refused there.
//
// What may not stand between the two is a link. Windows follows every
// component of a path but the last, so with a junction at `<worktree>/.tools`
// the Lstat of `<worktree>/.tools/godot` asks about the junction's target:
// absent there reads as "ours to fill", and the MkdirAll and junction.Create
// that follow write into whatever the junction points at -- outside the
// worktree, and at a path no sweep of ours ever looks at. Measured on
// 2026-09-08 without this check: exit 0, no output, and a junction standing at
// `<main>/elsewhere/godot`.
//
// Mode().IsDir() is the test, for the reason and the measurement written at
// standsInside. A component Lstat cannot read at all is refused with the same
// statement -- it is not *shown* to be a plain directory, and the message
// says no more than that.
func parentsPlainOrAbsent(dir, relative string) (string, bool) {
	// mirrorcfg hands out cleaned, slash-separated paths, so this split has no
	// empty component and no trailing one.
	components := strings.Split(relative, "/")
	path := dir
	for _, component := range components[:len(components)-1] {
		path = filepath.Join(path, component)
		info, err := os.Lstat(path)
		if os.IsNotExist(err) {
			// Nothing below an absent component exists either, so there is no
			// further component to look at.
			return "", true
		}
		if err != nil || !info.Mode().IsDir() {
			return path, false
		}
	}
	return "", true
}

// leadsInto says whether a link target sits inside the main checkout.
//
// The comparison is by what the two paths open and never by how they are
// spelled. A stored target's trailing separator depends on who made the
// junction -- measured on 2026-09-07, junction.Create writes `\??\C:\dir\` and
// `mklink /J` writes `\??\C:\dir`, and both resolve -- so a text comparison
// against an absolute path would miss exactly the hand-made links this
// repository's worktrees already carry.
//
// Walking up rather than testing the directory itself, because a configured
// path is a path and not only a name: `.ultraloom/vendor` names a target two
// levels below the main checkout.
func leadsInto(target, main string) bool {
	for path := stripNTPrefix(target); ; path = filepath.Dir(path) {
		if sameDir(path, main) {
			return true
		}
		if parent := filepath.Dir(path); parent == path {
			return false
		}
	}
}

// stripNTPrefix turns a stored reparse target into a path Go can stat.
//
// One of the two prefixes comes off if it is there, and a target carrying
// neither passes through untouched -- junction.Target promises the substitute
// name as the filesystem stored it and nothing about which form that is. The
// Clean is what takes the trailing separator off, so no caller has to know
// whether one was stored.
func stripNTPrefix(target string) string {
	for _, prefix := range ntPrefixes {
		if rest, found := strings.CutPrefix(target, prefix); found {
			return filepath.Clean(rest)
		}
	}
	return filepath.Clean(target)
}

// sameDir compares two paths by what they open, not by how they are spelled.
// A path that cannot be stat'ed is not the directory in question: os.Stat
// fails on "" as it does on anything else absent, so nothing here needs a
// guard against the empty string.
func sameDir(a, b string) bool {
	left, err := os.Stat(a)
	if err != nil {
		return false
	}
	right, err := os.Stat(b)
	if err != nil {
		return false
	}
	return os.SameFile(left, right)
}
